### PR TITLE
Remove global app logic from Window.__init__ 

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This follows on #2056 (and includes its as well) and completes the removal of application-creation logic from the `Window` class, along with  the `_napari_global_logo` and related class attributes on the Viewer.

@Czaki, you [expressed an opinion](https://github.com/napari/napari/pull/2056#discussion_r550945325) that all of this stuff should not be removed from `Viewer`, but i have a hard time understanding why global `QApplication` stuff needs to be there, and it makes our Viewer and Window objects clunkier.  I'd suggest that you just call `get_app` prior to instantiating the first Viewer in your application (essentially "claiming" the application as your own).

Please have a look at this PR and tell us if you are unable to accomplish what you need using some combination of  `get_app`, `set_app_id`,  and `viewer.window._qt_window.setWindowIcon()`.  If not, please give me an example that I can use to understand the limitations.  Thanks

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/_qt/qt_event_loop.py
Comment (Czaki) [inline napari/_qt/qt_event_loop.py]: This line calls windows specyfic code which controls gruping on windows. I do not understand why this is called on different level than `setApplicationName`

File: napari/_qt/qt_event_loop.py
Comment (Czaki) [inline napari/_qt/qt_event_loop.py]: calling these here change global behaviour. Now, if qt loop is created ealier, for example with `%gui qt` then application will not have set these properties.

**Review by Czaki — CHANGES_REQUESTED** [review]
> @Czaki, you [expressed an opinion](https://github.com/napari/napari/pull/2056#discussion_r550945325) that all of this stuff should not be removed from `Viewer`, but i have a hard time understanding why global `QApplication` stuff needs to be there, and it makes our Viewer and Window objects clunkier. I'd suggest that you just call `get_app` prior to instantiating the first Viewer in your application (essentially "claiming" the application as your own).


I only want to have possibility to prevent calling functions like `setWindowIcon`, `setApplicationName` and  `set_app_id` to do this on my own (maybe using same function with different parameters). Previous implementation, always calls these functions on create of viewer. Then I add logic which allows to prevent these calls. 

If You decide to break behaviour and call these functions only when creating Qapplication instance then it will not break my code.

File: napari/_qt/qt_event_loop.py
Comment (tlambert03) [reply napari/_qt/qt_event_loop.py]: - that is by design.  If ipython created the application, then we _don't want_ to set those properties.
- this is how it is with `applicationName` [already is in master](url)
- all of those other properties besides `applicationName` are new anyway, so they're not being set in master as it is.

File: napari/_qt/qt_event_loop.py
Comment (tlambert03) [reply napari/_qt/qt_event_loop.py]: I would prefer it to be called only when the app is created as well, I only put it there for your convenience, since it's more similar to the behavior in master.  I'll make it stricter

**Comment by tlambert03** [comment]
> I only want to have possibility to prevent calling functions like setWindowIcon, setApplicationName and set_app_id to do this on my own

put yourself in the position of napari, can you not see the argument for removing these global app specific arguments from the (non-global) window?  _why_ don't you want to have to call these functions?

> If You decide to break behaviour  ...

what behavior do you see this as breaking?

File: napari/_qt/qt_event_loop.py
Comment (Czaki) [reply napari/_qt/qt_event_loop.py]: `app.setWindowIcon(QIcon(icon))` is called on all Viewer creation (if is not blocked by class variable) on master and only on creation application in this PR` 

https://github.com/napari/napari/blob/35c9fa19a58d0b6245cdfb186a2e085612dfc729/napari/_qt/qt_main_window.py#L125-L127

Comment (tlambert03) [reply napari/_qt/qt_event_loop.py]: Correct!

Comment (tlambert03) [reply napari/_qt/qt_event_loop.py]: (That's the point)

Comment (tlambert03) [reply napari/_qt/qt_event_loop.py]: I guess what I'm asking is, rather than pointing out what this PR is doing (which I'm aware of), can you tell us whether there's anything you're unable to do in your app with this

**Comment by Czaki** [comment]
> put yourself in the position of napari, can you not see the argument for removing these global app specific arguments from the (non-global) window? _why_ don't you want to have to call these functions?

I wany to call these functions with my own set of parameters, not napari default one.

**Comment by tlambert03** [comment]
I'm still not following.  What's stopping you from calling `get_app` with your own set of parameters?

**Comment by Czaki** [comment]
> I'm still not following. What's stopping you from calling `get_app` with your own set of parameters?

Nothing. My wories in #2056 are conected with previous behaviour of seting icon and application idetifier on all `Viewer` constructor call.

**Comment by tlambert03** [comment]
> Nothing. My wories in #2056 are conected with previous behaviour of seting icon and application idetifier on all `Viewer` constructor call.

great!  sounds like you're ok with this then.

File: napari/_qt/qt_event_loop.py
Comment (sofroniewn) [inline napari/_qt/qt_event_loop.py:55]: If the app already exists has the name `napari`, and then you call it with a setting here like `my_custom_name`, will it get renamed as such? If so then I might change `if creating for the first time`, if not what are your thoughts about emitting a warning if the user tries to set it and they are unable to?

This applies to all of these similar input args

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [inline napari/_qt/qt_main_window.py:83]: I love that this is getting removed from this file!

File: napari/_qt/qt_event_loop.py
Comment (sofroniewn) [inline napari/_qt/qt_event_loop.py:167]: Can the icon path here somehow get read from the `app`? I guess not really needed as you're calling `get_app` with defaults, but it just seems safer to only have this set once - i.e. maybe we store it as `app._icon_path`. Up to you, it's just a rough idea

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [inline napari/_qt/qt_main_window.py:43]: Similarly here, I'm wondering why we're importing and storing the `NAPARI_ICON_PATH`, instead of maybe getting it from `app._icon_path` (assuming we add something like that in `get_app`, and then pass that to this function. While it's not necessary I think I would find that a bit simpler from a code perspective.

**Review by sofroniewn — APPROVED** [review]
I love this @tlambert03, I think it cleans things up a lot. I've left some minor comments, curious what you think of them.

It sounds like the current behaviour is flexible enough to meet the needs of @Czaki, and that it's actually an improvement on what we had before which is nice. 

We'll make some `rc` before the actual release so you'll have time to double check all this after we merge, but do let us know if we end up breaking something for you as we'll try and make sure you can still do what you want.

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py:43]: I did that as a convenience for @Czaki... If he wants to, in his application, he can patch `napari._qt.qt_main_window._QtMainWindow._window_icon` to whatever string he wants to, and then that will be the icon used (rather than him needing to call `viewer.window._qt_window.setWindowIcon` after every instantiation).  (this gives it a safe namespace that can be modified without changing our internal `NAPARI_ICON_PATH`)

File: napari/_qt/qt_event_loop.py
Comment (tlambert03) [reply napari/_qt/qt_event_loop.py:167]: I actually looked into that a bit.  I was unable to find a way to extract the fileName from a [`QIcon`](https://doc.qt.io/qt-5/qicon.html), and I'd rather not set attributes on the application object (we do it with `_existed` but I want to get rid of that)...  In any case, I think the splash window and the application window icon are different enough that we don't necessarily always want one to be the other.

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py:43]: Ok I get it. That's fine. Maybe you can add a comment here like
```
# This was added so that someone can patch `napari._qt.qt_main_window._QtMainWindow._window_icon` to
# whatever string they want, and that icon will always be used, and so
# `viewer.window._qt_window.setWindowIcon` won't need to be called after every instantiation.

File: napari/_qt/qt_event_loop.py
Comment (sofroniewn) [reply napari/_qt/qt_event_loop.py:167]: yeah, ok, that's fine too!

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py:43]: added

File: napari/_qt/qt_event_loop.py
Comment (tlambert03) [reply napari/_qt/qt_event_loop.py:55]: nothing gets changed on the second call to get_app.  I added a warning. I also made it so that those kwargs are "all or nothing".  (i.e. if none of them are set, then it uses the napari defaults, if any are provided, it only uses the provided values)

Comment (sofroniewn) [reply napari/_qt/qt_event_loop.py:55]: Great, thanks!

**Comment by sofroniewn** [comment]
Ok, thanks @tlambert03 I think this is good to go. I will merge now, as I said above we can still adjust during our rc process!

**Pull request merged**
